### PR TITLE
[4.x] Menu Items with icon only

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue #34639 .

### Summary of Changes
Enable usage of icons only in frontend menus. The user can now define an icon (for example icon-home or fas fa-user ) and build a menu using only icons without text.

NOTE: This PR does NOT care for code which seems copied for many years from one menuType to the other and results in lots of  code. 

The PR
- Adds a new field to all  menu types: menu_icon_css
- Uses this in all menuTypes in the same way - this means that the menuItems which now are different in some way now have the same behaviour.

A menuItem can be just the title
or be an image with or without the title 
or be an icon with or without the title (this is new).

If a menu Item is only an icon, the menuitem title is added visually hidden (a11y)
If a menuItem is not hidden, the icon gets the aria-hidden="true"

### Testing Instructions
Build menus for example with blog sample data.
Change menu Items in main menu and blog menu (dropdown) and play with params: 

### Actual result BEFORE applying this Pull Request
Before patch- see #34639 .

### Expected result AFTER applying this Pull Request
Example 

grafik

### Documentation Changes Required
yes

### CODE REVIEW

**Comment by dgrammatiko** [comment]
Imposing loading the font awesome icon in the front end due to some option on the back end is a **terrible idea**, (also this will never work on any template that doesn't load any icon fonts).
Either:
- use Jlayouts for the markup
- use svg icons, it's 2021 and we should stop using Icon fonts in the front end!

**Comment by brianteeman** [comment]
if you want to use an svg you can already do this.
this pr does not impose font awesome at all. it will work with any icon font the template wants to use
obviously the template would have to support an icon font for it to work

**Comment by brianteeman** [comment]
I have tested this and it does exactly what it says it will do. This is a good addition

**Comment by chmst** [comment]
@dgrammatiko we don't add icon fonts. 

The issue was: users now add icons via the class for the link (see the issue #30793, #32471 and others ). 
This causes a11y error and makes it impossible to build a  menu with icons only. 
This PR only let the user separate the icon from menu title. 

I agree that this should be a Layout - at the moment the code is repeated in different places. But this is not in scope of this PR

**Comment by brianteeman** [comment]
To avoid confusion with people trying to type a path to an icon etc why not label it icon class

image

**Comment by chmst** [comment]
@brianteeman thank you, will do in the next step. There is another language string 
COM_MENUS_ITEM_FIELD_ICON_TITLE_LABEL="Link Icon Class"
Maybe we can use this one in all layouts.

**Comment by ChristineWk** [comment]
> 
> 
> I have tested this and it does exactly what it says it will do. This is a good addition

confirm above.

> 
> 
> @dgrammatiko we don't add icon fonts.
> 
> The issue was: users now add icons via the class for the link (see the issue #30793, #32471 and others ).
> This causes a11y error and makes it impossible to build a menu with icons only.
> This PR only let the user separate the icon from menu title.
> 
> I agree that this should be a Layout - at the moment the code is repeated in different places. But this is not in scope of this PR

tested on 4.0.0-rc4-dev and confirm:
a) icon, e.g. fa fa-home & no title = OK.
b) icon, e.g. fa fa-search & no title = OK, but changes the font, as it is class for a link. (it's known, see above issues)

Above tests, shows only a summary.

**Comment by chmst** [comment]
@ChristineWk thanks for testing. 
Your case b) If the user enters an icon class into the link - this is a problem (the same as in blog layout, if a user says num_columns = 3 and enters columns-2 into the class field.  No clue what to do then.

@brianteeman your opinion, please: We cannot avoid that a user adds an icon class AND an image. I had the image with higher proirity. 

These language keys exist already
COM_MENUS_ITEM_FIELD_ICON_TITLE_DESC="If an icon class is entered, it takes precedence over the link image."
COM_MENUS_ITEM_FIELD_ICON_TITLE_LABEL="Link Icon Class"

**Comment by brianteeman** [comment]
> We cannot avoid that a user adds an icon class AND an image.

Why not? 
```
			<field
				name="menu_image"
				type="media"
				label="COM_MENUS_ITEM_FIELD_MENU_IMAGE_LABEL"
				showon="menu-anchor_icon:"
			/>
```

Also for consistency please rename the new fields menu_* instead of menu-*

**Comment by ChristineWk** [comment]
With Patch got now (for menus) in frontend:

`Notice: Undefined property: Joomla\CMS\Menu\MenuItem::$menu_icon in /example.org/templates/cassiopeia/html/mod_menu/dropdown-metismenu_component.php on line 44 link of menu items`

will try to load a new Nightly and try to remove the overrides I did before

**Comment by chmst** [comment]
@brianteeman the showon is possible of course, but li feel it more confusing than helpful, especially if a user changes his mind while he is building a menu. 

The more I dive into these scripts, the more I see how it is not consistent. For example - the menuItem title also is used as alt-text for the image. This is not really wrong, but also is not as clear as we have it now for other images. 

As @dgrammatiko says, these different menuIten type layouts should be improved, but not now in J4.0

File: modules/mod_menu/tmpl/default_component.php [hunk 1]
```diff
@@ -41,21 +41,38 @@
...
+		// The link is an icon
+		if ($itemParams->get('menu_text', 1))
+		{
+			// If the link text is to be displayed, the icon is added with aria-hidden
+			$linktype = '<span class="p-2 ' . $item->menu_icon . '" aria-hidden="true"></span>' . $item->title;
+		}
+		else
+		{
+			// If the icon itself is the link, it needs a visually hidden text
+			$linktype = '<span class="p-2 ' . $item->menu_icon . '" ></span><span class="visually-hidden">' . $item->title . '</span>';
```
Comment (Quy) [inline modules/mod_menu/tmpl/default_component.php]:
```suggestion
			$linktype = '<span class="p-2 ' . $item->menu_icon . '"></span><span class="visually-hidden">' . $item->title . '</span>';
```

Comment (Quy) [reply modules/mod_menu/tmpl/default_component.php]: Please change the other instances too.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 5a8291ca01f7969929b721cd9c822342666fe991<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34658">issues.joomla.org/tracker/joomla-cms/34658</a>.</sub>

**Comment by ChristineWk** [comment]
Still getting:

`Notice: Undefined property: Joomla\CMS\Menu\MenuItem::$menu_icon in /example.org/templates/cassiopeia/html/mod_menu/dropdown-metismenu_component.php on line 44 link of menu items`

Tried also with Prebuilt Package and then installed Nightly from today.

**Comment by chmst** [comment]
Thanks @ChristineWk! 
Don't know why this was not in the latest commit. It happened when there were old menuItems.

**Comment by ChristineWk** [comment]
Thanks @chmst 
Notice has disappeared & the font in the Search menu is also back to normal. 
But you can still see the title. Will keep trying. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34658">issues.joomla.org/tracker/joomla-cms/34658</a>.</sub>

**Comment by chmst** [comment]
I did not change the title. This is not in scope of this PR.

**Comment by N6REJ** [comment]
glad to see this becoming part of core.  I used to have to do work arounds to make this happen.

**Comment by Quy** [comment]
No translation for menu type URL: `COM_MENUS_ITEM_FIELD_ICON_LABEL`

File: modules/mod_menu/src/Helper/MenuHelper.php [hunk 2]
```diff
@@ -170,11 +170,11 @@ public static function getList(&$params)
 					// We prevent the double encoding because for some reason the $item is shared for menu modules and we get double encoding
 					// when the cause of that is found the argument should be removed
 					$item->title          = htmlspecialchars($item->title, ENT_COMPAT, 'UTF-8', false);
+					$item->menu_icon      = htmlspecialchars($itemParams->get('menu-icon', ''), ENT_COMPAT, 'UTF-8', false);
```
Comment (Quy) [inline modules/mod_menu/src/Helper/MenuHelper.php]:
```suggestion
					$item->menu_icon      = htmlspecialchars($itemParams->get('menu_icon', ''), ENT_COMPAT, 'UTF-8', false);
```

File: modules/mod_menu/tmpl/default_component.php [hunk 3]
```diff
@@ -41,21 +41,38 @@
...
+		// The link is an icon
+		if ($itemParams->get('menu_text', 1))
+		{
+			// If the link text is to be displayed, the icon is added with aria-hidden
+			$linktype = '<span class="p-2 ' . $item->menu_icon . '" aria-hidden="true"></span>' . $item->title;
+		}
+		else
+		{
+			// If the icon itself is the link, it needs a visually hidden text
+			$linktype = '<span class="p-2 ' . $item->menu_icon . '"></span><span class="visually-hidden">' . $item->title . '</span>';
```
Comment (Quy) [inline modules/mod_menu/tmpl/default_component.php]: Add `aria-hidden="true"`?

File: administrator/components/com_menus/forms/item_alias.xml [hunk 4]
```diff
@@ -44,6 +44,14 @@
 				name="menu-anchor_css"
 				type="text"
 				label="COM_MENUS_ITEM_FIELD_ANCHOR_CSS_LABEL"
+				validate="CssIdentifier"
+			/>
+
+			<field
+				name="menu_icon"
+				type="text"
+				label="COM_MENUS_ITEM_FIELD_ICON_TITLE_LABEL"
```
Comment (Quy) [inline administrator/components/com_menus/forms/item_alias.xml:53]: Add `COM_MENUS_ITEM_FIELD_ICON_TITLE_DESC`?

File: modules/mod_menu/tmpl/default_url.php [hunk 5]
```diff
@@ -31,21 +31,38 @@
 
 $linktype = $item->title;
 
-if ($item->menu_image)
+if ($item->menu_image || $item->menu_icon )
```
Comment (Quy) [inline modules/mod_menu/tmpl/default_url.php]: Change to this syntax?

```
if ($item->menu_icon)
{
	
}
elseif ($item->menu_image)
{
	
}
```

File: administrator/components/com_menus/forms/item_heading.xml [hunk 6]
```diff
@@ -17,6 +17,13 @@
 				validate="CssIdentifier"
 			/>
 
+			<field
+				name="menu_icon"
```
Comment (Quy) [inline administrator/components/com_menus/forms/item_heading.xml]: Rename `menu_icon_css`?

File: administrator/components/com_menus/forms/item_alias.xml (hunk 4 above)
Comment (chmst) [reply administrator/components/com_menus/forms/item_alias.xml:53]: This description is not correct for what I changed. Unfortunately the whole code for icons is not consistent, so this message is wrong.

**Comment by mariantanase** [comment]
I have tested this item :white_check_mark: successfully on #34658

File: administrator/components/com_menus/forms/item_alias.xml (hunk 4 above)
Comment (Quy) [reply administrator/components/com_menus/forms/item_alias.xml:53]: Why is it not correct? Reading the code, it appears to be doing exactly this.

> COM_MENUS_ITEM_FIELD_ICON_TITLE_DESC="If an icon class is entered, it takes precedence over the link image."

Comment (chmst) [reply administrator/components/com_menus/forms/item_alias.xml:53]: It is the other way around. If an image is entered, it takes priority. From my experience with sites, an image as mneu item is more common than only an icon. 
Of course we can change this, matching the older description. I have. 

It also would be possible to enable both - image and icon. But this needs a refactoring of the code.

Comment (Quy) [reply administrator/components/com_menus/forms/item_alias.xml:53]: Yes, you are right! I don't see a use case to enable both at the same time.

File: modules/mod_menu/tmpl/default_component.php (hunk 3 above)
Comment (chmst) [reply modules/mod_menu/tmpl/default_component.php]: thx. done

File: administrator/components/com_menus/forms/item_alias.xml (hunk 4 above)
Comment (chmst) [reply administrator/components/com_menus/forms/item_alias.xml:53]: added description and changed preference

File: administrator/components/com_menus/forms/item_heading.xml (hunk 6 above)
Comment (chmst) [reply administrator/components/com_menus/forms/item_heading.xml]: done

**Comment by chmst** [comment]
Changed as reqired. Thank you @Quy!

Thanks also @mariantanase for testing. I have improved the code with suggestions. It would be great if you could repeat the test (when @Quy has approved the changes)

File: administrator/components/com_menus/forms/itemadmin_alias.xml [hunk 7]
```diff
@@ -27,6 +27,13 @@
 				label="COM_MENUS_ITEM_FIELD_ANCHOR_TITLE_LABEL"
 			/>
 
+
+			<field
```
Comment (Quy) [inline administrator/components/com_menus/forms/itemadmin_alias.xml]:
```suggestion
			/>

			<field
```

File: modules/mod_menu/tmpl/default_component.php [hunk 8]
```diff
@@ -41,17 +41,31 @@
...
 	else
 	{
-		$linktype = HTMLHelper::_('image', $item->menu_image, $item->title);
+		// If the icon itself is the link, it needs a visually hidden text
+		$linktype = '<span class="p-2 ' . $item->menu_icon . '" aria-hidden="true"></span><span class="visually-hidden">' . $item->title . '</span>';
 	}
+}
+elseif ($item->menu_image)
+{
+	// The link is an image, maybe with an own class
```
Comment (Quy) [inline modules/mod_menu/tmpl/default_component.php]:
```suggestion
	// The link is an image, maybe with its own class
```

File: modules/mod_menu/tmpl/default_heading.php [hunk 9]
```diff
@@ -15,17 +15,31 @@
...
 	else
 	{
-		$linktype = HTMLHelper::_('image', $item->menu_image, $item->title);
+		// If the icon itself is the link, it needs a visually hidden text
+		$linktype = '<span class="p-2 ' . $item->menu_icon . '" aria-hidden="true"></span><span class="visually-hidden">' . $item->title . '</span>';
 	}
+}
+elseif ($item->menu_image)
+{
+	// The link is an image, maybe with an own class
```
Comment (brianteeman) [inline modules/mod_menu/tmpl/default_heading.php]:
```suggestion
	// The link is an image, maybe with its own class
```

File: modules/mod_menu/tmpl/default_separator.php [hunk 10]
```diff
@@ -15,17 +15,31 @@
...
 	else
 	{
-		$linktype = HTMLHelper::_('image', $item->menu_image, $item->title);
+		// If the icon itself is the link, it needs a visually hidden text
+		$linktype = '<span class="p-2 ' . $item->menu_icon . '" aria-hidden="true"></span><span class="visually-hidden">' . $item->title . '</span>';
 	}
+}
+elseif ($item->menu_image)
+{
+	// The link is an image, maybe with an own class
```
Comment (brianteeman) [inline modules/mod_menu/tmpl/default_separator.php]:
```suggestion
	// The link is an image, maybe with its own class
```

File: modules/mod_menu/tmpl/default_url.php (hunk 5 above)
Comment (chmst) [reply modules/mod_menu/tmpl/default_url.php]: done, thx.

**Comment by mariantanase** [comment]
I have re-tested this item ✅ successfully on #34658 

Anyway, I think is better to say:
"If a **Link Icon Class** is entered, it takes precedence over the **Link Image**."
instead of:
"If an icon class is entered, it takes precedence over the link image."

Or, better that that, consider to add that text as a tooltip for "Link Icon Class", to reduce the vertical space between two fields.

**Comment by chmst** [comment]
Thanks for testing. At the moment no changes in language strings are possible https://github.com/joomla/joomla-cms/discussions/34685 . So you could open an issue for a later version.

**Comment by brianteeman** [comment]
__Could someone remove the "updates requested" label please

**Review by hans2103 — APPROVED** [review]

**Comment by hans2103** [comment]
I have tested this item :white_check_mark: successfully on 97fee991ed2e0240b2f7f1c2702c57d16c1a3b1f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34658">issues.joomla.org/tracker/joomla-cms/34658</a>.</sub>

**Comment by richard67** [comment]
@mariantanase I have altered your test result in the issue tracker so it is properly counted. Next time please us the "Test this" button in the issue tracker to submitt your test result. Making a comment with just a green check mark isn't conuted right by our tools. Thanks in advance, and thanks for testing.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34658">issues.joomla.org/tracker/joomla-cms/34658</a>.</sub>

**Comment by richard67** [comment]
@wilsonge @bembelimen Would you say this is a bug fix so it can go into 4.0.x, or is it a new feature so it has to go into 4.1? I've set the RLDQ label because I think it needs that decision.

**Comment by bembelimen** [comment]
4.1 from my point of view.

**Comment by chmst** [comment]
It is a new feature from my point of view

**Comment by bembelimen** [comment]
Thx

**Pull request merged**